### PR TITLE
Add device routes

### PR DESCRIPTION
**Pull request by nahiluhmot**
@djosborne Add RESTful routes for the frontend.

@Daniel0524 @AdamEdgett I recommend you guys take a peek at the [controller spec](https://github.com/cappity-cappity-capstone/backend/blob/eaabcaf5a3a710d8dbd8612cd57290be1a0a48c0/spec/app/cappy/controllers/devices_spec.rb) to get a sense for the HTTP status codes/responses I went with. For a tl;dr, here's the spec output

```
Cappy::Controllers::Devices
  GET /devices/
    returns a list of every device
  POST /devices/
    when invalid data is POSTed
      returns a "400 Client Error"
    when valid data is POSTed
      but that device already exists
        returns a "409 Conflict"
      and that device does not exist
        returns a "201 Created"
  GET /devices/DEVICE_ID/
    when the DEVICE_ID does not exist
      returns a "404 Not Found"
    when the DEVICE_ID exists
      returns that device
  PUT /devices/DEVICE_ID/
    when the DEVICE_ID does not exist
      returns a "404 Not Found"
    when the DEVICE_ID exists
      but invalid data is PUT
        returns a "400 Client Error"
      and valid data is PUT
        updates that device
  DELETE /devices/DEVICE_ID/
    when the DEVICE_ID does not exist
      returns a "404 Not Found"
    when the DEVICE_ID exists
      deletes that device
```

When not specified, the HTTP status is either `200` or `204`

### CODE REVIEW

File: app/cappy/controllers/base.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+          status 400
+        when Errors::NoSuchDevice
+          status 404
+        when Errors::DuplicationError
+          status 409
+        else
+          status 500
+        end
+
+        { class: err.class, message: err.message }.to_json
```
Comment (nahiluhmot) [inline app/cappy/controllers/base.rb]: Little nervous about this, but on the other hand we're not shipping to prod :beer:

**Comment by nahiluhmot** [comment]
@tlunter Would be sweet if you peeked at this too.

**Comment by nahiluhmot** [comment]
The web server can be started via `bundle exec rake web`

File: app/cappy/services/devices.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      end
+
+      def get_device(device_id)
+        Models::Device.find_by(device_id: device_id).tap do |device|
+          fail Errors::NoSuchDevice, device_id unless device
+        end
+      end
+
+      def wrap_active_record_errors(&block)
+        block.call
```
Comment (tlunter) [inline app/cappy/services/devices.rb]: Why not just a `yield`?

File: unicorn.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# Set the application directory
+APP_DIR = File.expand_path(File.dirname(__FILE__))
+TMP_DIR = File.join('/', 'tmp')
```
Comment (tlunter) [inline unicorn.rb]: `Dir.tmpdir` exists which respects `ENV['TMPDIR']`

Comment (Daniel0524) [inline unicorn.rb]: I would like to ride a unicorn.

Comment (Daniel0524) [inline unicorn.rb]:
```
                                                    /
                                                  .7
                                       \       , //
                                       |\.--._/|//
                                      /\ ) ) ).'/
                                     /(  \  // /
                                    /(   J`((_/ \
                                   / ) | _\     /
                                  /|)  \  eJ    L
                                 |  \ L \   L   L
                                /  \  J  `. J   L
                                |  )   L   \/   \
                               /  \    J   (\   /
             _....___         |  \      \   \```
      ,.._.-'        '''--...-||\     -. \   \
    .'.=.'                    `         `.\ [ Y
   /   /                                  \]  J
  Y / Y                                    Y   L
  | | |          \                         |   L
  | | |           Y                        A  J
  |   I           |                       /I\ /
  |    \          I             \        ( |]/|
  J     \         /._           /        -tI/ |
   L     )       /   /'-------'J           `'-:.
   J   .'      ,'  ,' ,     \   `'-.__          \
    \ T      ,'  ,'   )\    /|        ';'---7   /
     \|    ,'L  Y...-' / _.' /         \   /   /
      J   Y  |  J    .'-'   /         ,--.(   /
       L  |  J   L -'     .'         /  |    /\
       |  J.  L  J     .-;.-/       |    \ .' /
       J   L`-J   L____,.-'`        |  _.-'   |
        L  J   L  J                  ``  J    |
        J   L  |   L                     J    |
         L  J  L    \                    L    \
         |   L  ) _.'\                    ) _.'\
         L    \('`    \                  ('`    \
          ) _.'\`-....'                   `-....'
         ('`    \
          `-.___/  
```

File: app/cappy/services/devices.rb (hunk 2 above)
Comment (Daniel0524) [inline app/cappy/services/devices.rb]:
```

                                                                                
                                                                                
.. .                                      . ............................  ..    
 $  .......................................                                  O  
.7IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII7 .
.IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII7.D
7+7IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII?+.
.:I7IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII.  
...IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII7I.:. 
  ~ 7IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII7.7   
   ..7IIIIIIIIIIIIIII7I7777777777777777777777IIIIIIII77777IIIIIIIIIIIIIIII.8    
   .:.IIIIIIIIIIIIII.......................................7IIIIIIIIIIIII.I.    
     7.IIIIIIIIIIIIII..  .II..77.I+.IIII7.77  ..III?..   .IIIIIIIIIIIIII.8..    
     .D.IIIIIIIIIIIII?.  ..77=I..7+.I7....I7  ..I..I+  ..=IIIIIIIIIIIII,D.      
      .7.7IIIIIIIIIIII~.. ..II...I+.II7I..I7  ..I..77...:IIIIIIIIIIII7.Z.       
      ..8.IIIIIIIIIIIII?.   I7...I+.I7....I7  ..I..7I..,IIIIIIIIIIII7,D..       
        .?+IIIIIIIIIIIII.. .I7...I+.I7....II....I.I7...7IIIIIIIIIIII?=..        
          ,~7IIIIIIIIIIII...?I...I+.?IIII.III?.,II=...IIIIIIIIIIIII7.           
        . .??IIIIIIIIIIIII........... ...............I7IIIIIIIIIII?. .          
           .,7IIIIIIIIIIIII..                      .7IIIIIIIIIIII7,.            
           ...77IIIIIIIIIII7.                    ..7IIIIIIIIIIII7...            
             ..IIIIIIIIIIIII7...                ..7IIIIIIIIIII7I..              
             .:.7IIIIIIIIII7I7..                .7IIIIIIIIIIII7.:.              
              ...7IIIIIIIIIIIII.              ..77IIIIIIIIIII7.I                
              ..:.7IIIIIIIIIIII7..            .IIIIIIIIIIIIII.,.                
                 D.IIIIIIIIIIIII+.          ..IIIIIIIIIIIII7.8 .                
                  O.7IIIIIIIIIII7:..        .=IIIIIIIIIIII7,8                   
                 . D.IIIIIIIIIIIII..      ..,IIIIIIIIIIIII.7.                   
                   .O~IIIIIIIIIIIII:..   . .IIIIIIIIIIIII,D.                    
                    .+?IIIIIIIIIIIII..  ..:IIIIIIIIIIIII=$                      
                    ..,~IIIIIIIIIIII7.  .,IIIIIIIIIIIII?= .                     
                       ??IIIIIIIIIIII7...IIIIIIIIIIII7~..                       
                      ...IIIIIIIIIIIIIIIIIIIIIIIIIIII?+                         
                      ....IIIIIIIIIIIIIIIIIIIIIIIIIII. ..                       
                       . =.7IIIIIIIIIIIIIIIIIIIIIII7..                          
                         .Z.IIIIIIIIIIIIIIIIIIIIII7.~                           
                          .=.IIIIIIIIIIIIIIIIIIII7.7.                           
                           .Z.IIIIIIIIIIIIIIIIIII.:..                           
                             D,IIIIIIIIIIIIIII7I.I.                             
                             .O.IIIIIIIIIIIIIII.8 .                             
                              .D,IIIIIIIIIII7I.D                                
                             .  O~IIIIIIIIIII.8..                               
                                ..7IIIIIII7I=7  .                               
                                  .?77IIII7I~                                   
                                   ,7IIIII~.                                    
                                  ..,7III?=.                                    
                                   ..D...D..   
```

**Comment by nahiluhmot** [comment]
@tlunter Responded to ya feedback.

**Comment by tlunter** [comment]
:+1:

**Pull request merged**